rpc/src/client/transport/websocket.rs

```diff
@@ -0,0 +1,701 @@
+        let driver_handle = tokio::spawn(async move { driver.run().await });
+        Ok(Self {
+            host,
+            port,
+            driver_handle,
+            cmd_tx,
+            terminate_tx,
+        })
+    }
+
+    async fn send_cmd(&mut self, cmd: WebSocketDriverCmd) -> Result<()> {
+        self.cmd_tx.send(cmd).await.map_err(|e| {
+            Error::client_internal_error(format!("failed to send command to client driver: {}", e))
+        })
+    }
+}
+
+#[async_trait]
+impl SubscriptionClient for WebSocketSubscriptionClient {
+    async fn subscribe(&mut self, query: String) -> Result<Subscription> {
+        let (event_tx, event_rx) = unbounded();
+        let (result_tx, mut result_rx) = unbounded::<Result<()>>();
+        // NB: We assume here that the wrapper generates a unique ID for our
+        // subscription.
+        let req = request::Wrapper::new(subscribe::Request::new(query.clone()));
+        let id: SubscriptionId = req.id().clone().try_into()?;
+        self.send_cmd(WebSocketDriverCmd::Subscribe {
+            req,
+            event_tx,
+            result_tx,
+        })
+        .await?;
+        // Wait to make sure our subscription request went through
+        // successfully.
+        result_rx.recv().await.ok_or_else(|| {
+            Error::client_internal_error("failed to hear back from WebSocket driver".to_string())
+        })??;
+        Ok(Subscription::new(
+            id,
+            query,
+            event_rx,
+            self.terminate_tx.clone(),
+        ))
+    }
+}
+
+#[async_trait]
+impl ClosableClient for WebSocketSubscriptionClient {
+    /// Attempt to gracefully close the WebSocket connection.
+    async fn close(mut self) -> Result<()> {
+        self.cmd_tx.send(WebSocketDriverCmd::Close).await?;
+        self.driver_handle.await.map_err(|e| {
+            Error::client_internal_error(format!(
+                "failed while waiting for WebSocket driver task to terminate: {}",
+                e
+            ))
+        })?
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct GenericJSONResponse(serde_json::Value);
+
+impl Response for GenericJSONResponse {}
+
+#[derive(Debug)]
+struct WebSocketSubscriptionDriver {
+    stream: WebSocketStream<TokioAdapter<TcpStream>>,
+    router: SubscriptionRouter,
+    cmd_rx: ChannelRx<WebSocketDriverCmd>,
+    terminate_rx: ChannelRx<TerminateSubscription>,
+}
+
+impl WebSocketSubscriptionDriver {
+    fn new(
+        stream: WebSocketStream<TokioAdapter<TcpStream>>,
+        cmd_rx: ChannelRx<WebSocketDriverCmd>,
+        terminate_rx: ChannelRx<TerminateSubscription>,
+    ) -> Self {
+        Self {
+            stream,
+            router: SubscriptionRouter::default(),
+            cmd_rx,
+            terminate_rx,
+        }
+    }
+
+    async fn run(mut self) -> Result<()> {
+        // TODO(thane): Should this loop initiate a keepalive (ping) to the
```

Comment:
I believe so. At least the [Go client does this](https://github.com/tendermint/tendermint/blob/master/rpc/jsonrpc/client/ws_client.go#L421) and I think it's proper behaviour for a websocket client